Source/Private/ASRectMap.mm

```diff
@@ -0,0 +1,83 @@
+//
+//  ASRectMap.mm
+//  Texture
+//
+//  Copyright (c) 2014-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the /ASDK-Licenses directory of this source tree. An additional
+//  grant of patent rights can be found in the PATENTS file in the same directory.
+//
+//  Modifications to this file made after 4/13/2017 are: Copyright (c) 2017-present,
+//  Pinterest, Inc.  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import "ASRectMap.h"
+#import "ASObjectDescriptionHelpers.h"
+#import <UIKit/UIGeometry.h>
+#import <unordered_map>
+
+@implementation ASRectMap {
+  std::unordered_map<void *, CGRect> _map;
```

Comment:
We may should add a comment that we use a `void *` key here as arc would not be able to determine the ownership qualification.

Reply:
Could we not use `__unsafe_unretained id` as the key type here?

Reply:
Unfortunately we cannot. The `unordered_map` template cannot figure out the hash for the id key. We would have to write our own key like:

```
struct ASRectMapMapKey {
    __unsafe_unretained id obj;
    
    bool operator==(const ASRectMapMapKey &other) const
    {
        return [obj isEqual:other.obj];
    }
};

namespace std {
    template<> struct hash<ASRectMapMapKey>
    {
        size_t operator()(const ASRectMapMapKey &k) const
        {
            return [k.obj hash];
        }
    };
}
```

And we could use it via:
```
std::unordered_map<ASRectMapMapKey, CGRect> map;
map.insert({{@"key"}, CGRectMake(0, 0, 100, 100)});
NSLog(@"%@", NSStringFromRect(map.at({@"key"})));
```